feat: Support RUM context in trace headers.

fuzzybinary opened the pull request:
### What and why?

Add user id, account id, and session id to the `baggage` header in network requests.

The method for getting a user id and account id is different between iOS and Android, as iOS does not have FFI setup and Android does not broadcast context changes the way iOS does, so different implementation approaches needed to be used on each platform.

Web does not support these baggage headers at the moment, though web handles most of its own networking through XHR requests, not through our interceptors.

refs: RUM-8411

### Review checklist

- [ ] This pull request has appropriate unit and / or integration tests 
- [ ] This pull request references a Github or JIRA issue

fuzzybinary committed:
feat: Support RUM context in trace headers.

Add user id, account id, and session id to the `baggage` header in network requests.

The method for getting a user id and account id is different between iOS and Android, as iOS does not have FFI setup and Android does not broadcast context changes the way iOS does, so different implementation approaches needed to be used on each platform.

Web does not support these baggage headers at the moment, though web handles most of its own networking through XHR requests, not through our interceptors.

refs: RUM-8411

fuzzybinary committed:
Fix unit tests.

fuzzybinary committed:
Fix integration tests.

0xnm commented on packages/datadog_flutter_plugin/android/src/main/kotlin/com/datadoghq/flutter/DatadogSdkPlugin.kt:
```diff
@@ -38,6 +40,11 @@ class DatadogSdkPlugin : FlutterPlugin, MethodCallHandler {
+
+        fun getCoreContext(): DatadogContext? {
+            val sdkCore = Datadog.getInstance() as? InternalSdkCore
+            return sdkCore?.getDatadogContext()
```
Beware that this call is blocking. It will add a request to the context queue and wait until it is processed.

fuzzybinary replied:
Do I have another option?

0xnm replied:
I don't think so. That is why implementing caching is very beneficial.

0xnm commented on packages/datadog_flutter_plugin/jnigen.yaml:
```diff
@@ -18,4 +18,7 @@ classes:
+  - 'com.datadoghq.flutter.DatadogSdkPlugin'
+  - 'com.datadog.android.api.context.DatadogContext'
+  - 'com.datadog.android.api.context.UserInfo'
+  - 'com.datadog.android.api.context.AccountInfo'
```
```suggestion
  - 'com.datadog.android.api.context.AccountInfo'

```

0xnm commented on packages/datadog_flutter_plugin/lib/src/datadog_sdk_method_channel.dart:
```diff
@@ -8,13 +8,29 @@ import 'package:flutter/foundation.dart';
+    if (Platform.isIOS) {
+      return _cachedContext;
+    } else if (Platform.isAndroid) {
+      return AndroidDatadogFlutterPlugin.getContext();
```
we can probably also cache and invalidate it when `setUser`/`setAccount`/`addExtraInfo` calls are made

fuzzybinary replied:
That assumes those are the only ways for someone to set the user / account / extraInfo. It's unfortunately a bad assumption in hybrid apps.

fuzzybinary replied:
Unless you're saying Android should cache them?

0xnm replied:
No, I'm talking about the Flutter methods. Android won't cache anything here, because `DatadogContext` includes many properties.

0xnm commented on packages/datadog_grpc_interceptor/test/datadog_grpc_interceptor_test.dart:
```diff
@@ -473,7 +486,12 @@ void main() {
     mockDatadog = DatadogSdkMock();
+    when(() => mockDatadog.platform).thenReturn(mockPlatform);
+
+    when(() => mockDatadog.platform).thenReturn(mockPlatform);
```
things are repeated twice here

0xnm commented on packages/datadog_flutter_plugin/test/tracing/tracing_headers_test.dart:
```diff
@@ -371,16 +603,129 @@ void main() {
+
+    test('existing baggage header adds new values', () {
+      // Given
+      String oldBaggage = 'test_value_1=1,test_value_2=string';
```
FYI @marco-saia-datadog recently also added few tests https://github.com/DataDog/dd-sdk-reactnative/pull/1016/files#diff-395eac9b3950188084fb673149f381453a4f7fee342e91ed57f36ae94f0c3718, so maybe it is worth checking them as well.

fuzzybinary replied:
I've updated this to do the same amount of validation, but I have pending decision. I'll hold off on merging this until I get some clarification.

fuzzybinary committed:
Update packages/datadog_flutter_plugin/jnigen.yaml

Co-authored-by: Nikita Ogorodnikov <4046447+0xnm@users.noreply.github.com>

fuzzybinary committed:
Perform more thorough validation of baggage headers

This may be removed in favor of not modifying baggage headers we get from users, and instead just making sure our baggage is properly encoded.

fuzzybinary committed:
Fix missing export.

fuzzybinary committed:
Add dio baggage header checks.

fuzzybinary committed:
Fix dio unit tests.

fuzzybinary committed:
Do not drop invalid baggage keys. Prevent percent encode loops.

fuzzybinary commented:
@0xnm this is ready for re-review. I do think we want to take the listener approach in the future but we'll stick with the context stall for now.

0xnm commented on packages/datadog_grpc_interceptor/test/datadog_grpc_interceptor_test.dart:
```diff
@@ -473,7 +486,11 @@ void main() {
     await server.serve(port: port);
 
+    mockPlatform = DatadogPlatformMock();
+    mockPlatform = DatadogPlatformMock();
```
minor: the same line is repeated twice

0xnm commented on packages/datadog_flutter_plugin/lib/src/tracing/baggage_helpers.dart:
```diff
@@ -0,0 +1,172 @@
+String _percentDecode(String value) {
+  // This looks for percent encoded values and replaces them with their decoded
+  // values, leaving stand alone percents.
+  final escapeRE = RegExp(r'(?:%[\da-fA-F]{2})+');
```
should this be compiled and stored in a class/static property? seems there is no need to create a new instance each time given it is stateless.

fuzzybinary replied:
Yeah, I can move it up to be a global of the file.

0xnm commented on packages/datadog_flutter_plugin/lib/src/tracing/baggage_helpers.dart:
```diff
@@ -0,0 +1,172 @@
+  // This looks for percent encoded values and replaces them with their decoded
+  // values, leaving stand alone percents.
+  final escapeRE = RegExp(r'(?:%[\da-fA-F]{2})+');
+  return value.replaceAllMapped(escapeRE, (m) => Uri.decodeComponent(m[0]!));
```
Is it safe to do `!` here? Should a check for non-empty `m` be made before?

fuzzybinary replied:
I'm pretty sure the `!` is safe here. The mapping function shouldn't be called unless there's a match, which guarantees that there's an item in the first portion of the match. `m[0]` should only return `null` if there was no match when calling `RegExp.match`.

fuzzybinary committed:
Apply suggestions from code review.

Try to fix flakey integration test.

fuzzybinary committed:
Actually propagate baggage on web.
